### PR TITLE
Adding support for grouped datasets

### PR DESCRIPTION
**Pull request by brimoor**
This PR adds support for *grouped datasets*, eg multiple camera view scenes.

### Example usage

```py
import fiftyone as fo
import fiftyone.utils.random as four
import fiftyone.zoo as foz
from fiftyone import ViewField as F

#
# Generate some test data
#

groups = ["left", "center", "right"]

d = foz.load_zoo_dataset("quickstart")
four.random_split(d, {g: 1 / len(groups) for g in groups})
filepaths = [d.match_tags(g).values("filepath") for g in groups]

#
# Here's how grouped datasets are defined
#

# A dataset has `media_type == "group"` if it has one or more group fields declared via this syntax
dataset = fo.Dataset()
dataset.add_group_field("group", default="center")

samples = []
for fps in zip(*filepaths):
    group = fo.Group()
    for name, filepath in zip(groups, fps):
        sample = fo.Sample(filepath=filepath, group=group.element(name))
        samples.append(sample)

dataset.add_samples(samples)
print(dataset)

# Datasets with a `GroupField` have media type 'group'
print(dataset.media_type)
# group

# The names of all slices are available via the `group_slices` property
print(dataset.group_slices)
# ['left', 'center', 'right']

# Each slice must have a homogeneous media type. This dict exposes that info
print(dataset.group_media_types)
# {'left': 'image', 'center': 'image', 'right': 'image'}

# This property contains the name of the `GroupField` of the dataset (if any)
print(dataset.group_field)
# group

# Grouped datasets store a default slice
print(dataset.default_group_slice)
# center

# `group_slice` determines what slice's data will be returned if you directly
# iterate over a grouped dataset
print(dataset.group_slice)
# center

print(dataset.first())
```

```
<Sample: {
    'id': '62db2ce147e9efc3615cd450',
    'media_type': 'image',
    'filepath': '/Users/Brian/fiftyone/quickstart/data/003344.jpg',
    'tags': BaseList([]),
    'metadata': None,
    'group': <Group: {'id': '62db2ce147e9efc3615cd346', 'name': 'center'}>,
}>
```

```py
# This can be customized in the current session
dataset.group_slice = "left"
```

```
<Sample: {
    'id': '62db2ce147e9efc3615cd44e',
    'media_type': 'image',
    'filepath': '/Users/Brian/fiftyone/quickstart/data/001599.jpg',
    'tags': BaseList([]),
    'metadata': None,
    'group': <Group: {'id': '62db2ce147e9efc3615cd346', 'name': 'left'}>,
}>
```

```py
# The groups of a collection can also be iterated over
for group in dataset.iter_groups(progress=True):
    pass

print(group)
```

```
{'left': <Sample: {
    'id': '62db2ce147e9efc3615cd590',
    'media_type': 'image',
    'filepath': '/Users/Brian/fiftyone/quickstart/data/001227.jpg',
    'tags': BaseList([]),
    'metadata': None,
    'group': <Group: {'id': '62db2ce147e9efc3615cd44a', 'name': 'left'}>,
}>, 'center': <Sample: {
    'id': '62db2ce147e9efc3615cd591',
    'media_type': 'image',
    'filepath': '/Users/Brian/fiftyone/quickstart/data/000191.jpg',
    'tags': BaseList([]),
    'metadata': None,
    'group': <Group: {'id': '62db2ce147e9efc3615cd44a', 'name': 'center'}>,
}>, 'right': <Sample: {
    'id': '62db2ce147e9efc3615cd592',
    'media_type': 'image',
    'filepath': '/Users/Brian/fiftyone/quickstart/data/000781.jpg',
    'tags': BaseList([]),
    'metadata': None,
    'group': <Group: {'id': '62db2ce147e9efc3615cd44a', 'name': 'right'}>,
}>}
```

```py
# Or a specific group can be retrieved by its ID
group_id = dataset.first().group.id
group = dataset.get_group(group_id)

# Arbitrary views into grouped datasets can be created and iterated over
for group in dataset[-1:].iter_groups(progress=True):
    pass

dataset.compute_metadata()

# When writing views, you can reference specific slices of a grouped
# collection via the "groups.<slice>.field.name" syntax
view = dataset.match(
    (F("groups.left.metadata.height") >= 640)
    & (F("groups.right.metadata.height") <= 480)
)
print(view)

#
# `select_group_slices()` can be used to create a non-grouped view that contains one
# or more slices worth of data from the grouped dataset
#

# Create an image collection containing only left camera images
left_view = dataset.select_group_slices("left")
print(left_view)

# Create an image collection containing left and right camera images
lr_view = dataset.select_group_slices(["left", "right"])
print(lr_view)

#
# Image collection containing all camera images
#
# If the dataset contained groups with different media types, this would raise an error, because
# sample collections must be composed of a single homogeneous data type
#
flat_view = dataset.select_group_slices()
print(flat_view)

# Grouped collections can be imported/exported in FiftyOneDataset format
dataset.export(
    export_dir="/tmp/groups",
    dataset_type=fo.types.FiftyOneDataset,
)

dataset2 = fo.Dataset.from_dir(
    dataset_dir="/tmp/groups",
    dataset_type=fo.types.FiftyOneDataset,
)
```

### CODE REVIEW

**Comment by brimoor** [comment]
@benjaminpkane Hmm okay thanks for your thoughts 👍  let's do a deep dive offline

File: fiftyone/core/fields.py [hunk 1]
```diff
@@ -689,6 +688,35 @@ def __str__(self):
...
+
+    @property
+    def _id(self):
+        return ObjectId(self.id)
+
+    def element(self, name):
+        return self.__class__(id=self.id, name=name)
+
+
+class GroupField(EmbeddedDocumentField):
```
Comment (benjaminpkane) [inline fiftyone/core/fields.py:710]: Is this required? Could `EmbeddedDocumentField` just be used. `GroupField` will be an extra class that the App will have to hardcode for the sidebar, etc.

Comment (brimoor) [reply fiftyone/core/fields.py:710]: k I can remove `GroupField` and just use `EmbeddedDocumentField`

Comment (brimoor) [reply fiftyone/core/fields.py]: Done!

**Review by benjaminpkane — APPROVED** [review]
LGTM

**Pull request merged**